[DVCSMP-3206] Adding Fix for Aeon Labs Panic Button

skt123 opened the pull request:
@jackchi @ShunmugaSundar 
Added a new DTH for **Aeon Panic Button** which doesn't have implementation for child devices as the device has a single button. Kept the child device implementation intact for **Aeon Key Fob** as it works similar to Aeon Minimote. 
Going ahead with this implementation, we might need to enroll already installed **Aeon Panic Button** devices to the new DTH as well. 
Please review.

skt123 committed:
[DVCSMP-3206]
Adding Fix for Aeon Labs Panic Button and Aeon Key Fob

jackchi commented on devicetypes/smartthings/aeon-panic-button.src/aeon-panic-button.groovy:
```diff
@@ -0,0 +1,128 @@
+ *
+ */
+metadata {
+    definition (name: "Aeon Panic Button", namespace: "smartthings", author: "SmartThings") {
```
Does this mean we need UI Metadata? How does it look on Samsung Connect?

skt123 replied:
Yes, we'll need to add new UI Metadata for Aeon Panic Button DTH.

workingmonk commented:
why does require separate DTH? can't we have logic in the code to not break into child devices if `numOfbutton==1`?

skt123 commented:
@workingmonk @jackchi 
Both the DTHs also has tile changes which depends on the number of buttons that the device has. 
`numOfbutton` variable is interpreted only when the message is parsed and it won't be available in tile section. Thus the tile addition `childDeviceTiles("outlets")` for Key Fob won't be valid for Panic Button.

Or the other solution could be to create a single child device for Aeon Panic Button, but again it would be a duplicate of the parent Panic Button device.

workingmonk commented:
how will this fix the issue for existing users?

skt123 commented:
@workingmonk @jackchi 
This approach won't work for the devices that are already installed. Even calling the updated() method won't work as the logic for Panic Button resides in a new DTH. The only solution is to pair the device again but that does not seem to be a feasible option for the users. 

If right now for LAUNCH, we just need the Panic Button, we can drop the child device implementation from Key Fob DTH and revert the changes. Both Key Fob and Panic Button would share the same DTH again and we would not be facing any issues as such for existing users.

However, in future we might need the child device to be implemented for Key Fob as well and the correct way would be to separate out the DTH for these two devices. Going ahead, we will have to think of a way to work this out.

workingmonk commented:
lets discuss this properly before we merge this one in. if the motivation was the launch, we can discuss this privately instead of the public comment board

jackchi reviewed (commented):
@bflorian @workingmonk Any decisions on how to approach the Aeon Panic Button? 
It seems like we have 2 options:
1) Introduce a new DTH or
2) Drop the child device implementation

bflorian commented:
@jackchi but if there are no child devices created then the {{childDevices()}} tile command should not generate anything. Is that not what happens? Are the DTHs otherwise the same?

skt123 commented:
@jackchi 
Removed separate DTH for Aeon Panic Button.
Added extra checks in Key Fob DTH itself to handle child device implementation for Key Fob ONLY.
Please review.

cc @workingmonk @bflorian

bflorian commented on devicetypes/smartthings/aeon-key-fob.src/aeon-key-fob.groovy:
```diff
@@ -80,16 +81,22 @@ def zwaveEvent(physicalgraph.zwave.commands.wakeupv1.WakeUpNotification cmd) {
-	if (!child) {
-		log.error "Child device $childDni not found"
+	def child
+	if(state.buttons==4)
```
It's a small point but can we make this `if (state.buttons > 1)`? That way if they come out with a two or six button device it will work as-is

bflorian commented on devicetypes/smartthings/aeon-key-fob.src/aeon-key-fob.groovy:
```diff
@@ -128,42 +135,51 @@ def configure() {
-	for (i in num) {
+private void createChildDevices() {
+	state.child_count = 1
+	for (i in 1..4) {
```
As noted before, can we isolate the button count to a model check in one place and make it a variable elsewhere?

bflorian reviewed (changes requested):
Would like to see the number of buttons set in one place. Otherwise looks good

skt123 commented:
@bflorian 
Made the required changes. Please review.
cc @jackchi @workingmonk

workingmonk commented:
did you mean to use `state.child_count` as a boolean? if yes, why is it being set using numbers?

skt123 commented:
`state.child_count` was suppose to behave as a boolean variable with possible values 0/1. However, I have updated the assigned values to true/false.

workingmonk commented:
couple of things:
1. this will create child devices for buttons that already have child devices. there is no longer a check for that
2. why use state.buttons when you can device.currentState("numberOfButtons") ?
1. state.child_count is really state.child_created (incorrect naming)

skt123 commented:
@workingmonk, child device were created just once as it was being taken care by variable `state.child_count`. However I have made few alterations:
1. use of `device.currentState("numberOfButtons")` instead of `state.buttons`
2. removed `state.child_count`, using `childDevices` variable instead as done in Aeon Minimote.

workingmonk commented on devicetypes/smartthings/aeon-key-fob.src/aeon-key-fob.groovy:
```diff
@@ -129,41 +139,50 @@ def configure() {
 		buttons = 1
 	}
-	createChildDevices(buttons)
+	else {
```
you don't need the else block if you kept the code to what it was. the initialized value of 4

workingmonk commented on devicetypes/smartthings/aeon-key-fob.src/aeon-key-fob.groovy:
```diff
@@ -129,41 +139,50 @@ def configure() {
 	state.oldLabel = device.label
-	for (i in num) {
+	Integer buttons = (device.currentState("numberOfButtons").value).toBigInteger()
+	for (i in 1..buttons) {
```
across the code block, if the numberOfButtons is encountered as null, set it to the number of buttons that Aeon Panic button has. as this is a old DTH, it precedes the days before numberOfButtons was added to the platform

skt123 commented:
Added null check for `numberOfButtons` in `buttonEvent()` method. 
Also removed else condition while initializing number of buttons.
